### ISSUE TITLE
Allow text font size to scale when zooming

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#3185]
## 🚀 Feature
For some applications I think it would be nice to allow for text font size to scale when zooming, in the way that point size does. In the vispy marker visual this is set by `scaling=True` https://github.com/vispy/vispy/blob/af2439895a13f187d840a476bd73ed480f2978f2/vispy/visuals/markers.py#L526

It doesn't look like the vispy text visual supports this https://github.com/vispy/vispy/blob/af2439895a13f187d840a476bd73ed480f2978f2/vispy/visuals/text/text.py#L281

We might need to add this upstream to vispy and then expose it as a bool on the napari text manager.

Note this was also requested a long long time ago in https://github.com/vispy/vispy/issues/1072

Examples where this might be very nice are like `python examples/annotate_segmentation_with_text.py` where the text is associated with a bounding box.

cc @kevinyamauchi @alisterburt

**Comment by alisterburt** [comment]
great idea - will try to look into how to do the text scaling!

**Comment by kevinyamauchi** [comment]
I agree this is a good idea. Maybe investigating this will give us some ideas for fixing the scaling of points with perspective.

**Comment by sofroniewn** [comment]
Great! Thanks both!!

**Comment by bauerdavid** [comment]
Is there a built in solution for this? Anyways, I managed to do it like this:
```python
def reset_text_size_callback(layer):
    def on_zoom(self, event):
        layer.text.size = max(1, event.value)
    return on_zoom
viewer.camera.events.zoom.connect(reset_text_size_callback(my_layer))
```
It's not "pixel perfect", but the change is unnoticable. Also you cannot set the relative size, but this could also be easily solved.

**Comment by TimMonko** [comment]
> Is there a built in solution for this?

Not yet, I'm pretty sure. This is a good thing that I'll add to the tracking issue (#7587) about overlay and visualization. At least on my screen, your solution seems about as good as I'd hope (I guess `text.size` is rounded to int, which is the problem)

@bauerdavid can you give a little more detail about where you would like to see this feature? I imagine on any layer text, like with shapes or points? And just to confirm, the goal is to have the font size adjust dynamically to look the same relative to the object (point, shape, etc) at different zoom levels? This makes plenty of sense. Also, I think for the last part about relative size, I think you are saying you want to set the overall 'size' that is then scaled at each zoom level? 

Here's a script that I used to visualize
```
import numpy as np
import napari

viewer = napari.Viewer()
viewer.open_sample('napari', 'astronaut')

points = np.array([[100, 100], [120, 120], [140, 140], [160, 160]])
text = {
    'string': ['Point 1', 'Point 2', 'Point 3', 'Point 4'],
    'size': 30,  # Initial size, but is immediately overwritten
    'anchor': 'center'
}

points_layer = viewer.add_points(points, size=10, text=text)

def reset_text_size_callback(layer):
    def on_zoom(event):
        layer.text.size = max(1, event.value)
    return on_zoom

viewer.camera.events.zoom.connect(reset_text_size_callback(points_layer))

napari.run()

```

**Comment by TimMonko** [comment]
@bauerdavid great, thanks for the extra insight. It's helpful to understand this is helpful.

There doesn't seem to be any progress on the Vispy side, so perhaps this is something we can add on the napari side. 

One definite change I'll test out is if we can use float instead of int for the font size, which gives much more flexibility to use smaller font size (instead of 1) --- especially important with very small layer scales, like 0.01. 



> And a slightly off topic question: is there a reason for not having a slider in the layer UI for changing text size?

I'll look into it, but I don't think so. If there is no reason, I will make a PR to add it (unless you would like to!)

**Comment by psobolewskiPhD** [comment]
>  is there a reason for not having a slider in the layer UI for changing text size?

I think it's because you can't actually make text using the GUI.
So it would be confusing?

**Comment by bauerdavid** [comment]
> One definite change I'll test out is if we can use float instead of int for the font size, which gives much more flexibility to use smaller font size (instead of 1) --- especially important with very small layer scales, like 0.01.

Sound great, thanks!


> I think it's because you can't actually make text using the GUI.
> So it would be confusing?

Good point. Do you think it would be possible to allow the user to manage the displayed text? What I'm thinking about is a text field, where they can provide the format text (the same used by `Layer.text`). Of course this would only work if 1) The objects have some features, and 2) the user knows that there are some features. I could also imagine some text completion-like help (like Google guessing your prompt), with the available layer features. But I can see that this is something very special, I'm not sure if this is a good idea... (And it is off topic here in this issue anyway.)

**Comment by TimMonko** [comment]
> Good point. Do you think it would be possible to allow the user to manage the displayed text? What I'm thinking about is a text field, where they can provide the format text (the same used by `Layer.text`). Of course this would only work if 1) The objects have some features, and 2) the user knows that there are some features. I could also imagine some text completion-like help (like Google guessing your prompt), with the available layer features. But I can see that this is something very special, I'm not sure if this is a good idea... (And it is off topic here in this issue anyway.)

Can you make a new issue for this! It's a feature I would like to see as well, even if the implementation will require some thought. I really don't like how many things are 'hidden' via code-only language -- it makes it really inaccessible to many people _and_ it seems a lot of plugins already use layer text, so exposing the ability to edit things in-GUI would be good.

**Comment by psobolewskiPhD** [comment]
I think the way to go would be to add a text related dock widget, accessible from the View menu.

**Comment by bauerdavid** [comment]
> Can you make a new issue for this! It's a feature I would like to see as well, even if the implementation will require some thought. I really don't like how many things are 'hidden' via code-only language -- it makes it really inaccessible to many people _and_ it seems a lot of plugins already use layer text, so exposing the ability to edit things in-GUI would be good.

> I think the way to go would be to add a text related dock widget, accessible from the View menu.

@psobolewskiPhD @TimMonko here you are: https://github.com/napari/napari/issues/7726
